### ISSUE TITLE
An example

### ISSUE THREAD

**Issue by Fruckin** [Zhou198/GPS#1]
Could you give an example of how to use GPSClassifier? The code has no example so I do not know how to use it. When I want to run gps, it always give me an error.

**Comment by Zhou198** [comment]
Hi, I've included an example within this repository for the GPS classifier. It's annotated with some comments for better understanding. Please refer to the corresponding Jupyter notebook for detailed guidance.

**Comment by zrli2000** [comment]
Thank you for your early relpy. But I have a new question: if i just want to preform a regular classification(with no outliers), take iris data set for example, it can not run:


import numpy as np
from sklearn.model_selection import train_test_split, ShuffleSplit
import pandas as pd
from GPS import GPSclassifier

columns = ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)','class']
df = pd.read_table("../datasets/data/iris/iris.data",delimiter=",",names=columns)
target = 'class'
features = columns[:-1]
df[target] = df[target].astype('category')
target_names = list(df[target].cat.categories)
# transform the class label to 1,2,3
target_to_num = {target: i+1 for i, target in enumerate(target_names)}
df[target] = df[target].map(target_to_num)

X = df[features].values
y = df[target].values
K= np.unique(y).shape[0]

temp_X, test_X, temp_y, test_y = train_test_split(X, y, test_size=0.7, stratify=y)
train_X, calib_X, train_y, calib_y = train_test_split(temp_X, temp_y, test_size=0.5, stratify=temp_y)
calibLabelSet = np.hstack([calib_y.reshape(-1, 1), calib_X])

candC1 = [0.1, 1, 10]
candSigqtl = [25, 50, 75]

gps = GPSclassifier(alpha=0.05, K=3, calibSet=calibLabelSet, candC1=candC1, candSigqtl=candSigqtl)
gps.train(train_X, train_y)

**Comment by Fruckin** [comment]
Thank you for your early relpy. But I have two new questions: 
1. if in the train and valid process, there must have outlier data in?
2. if I just want to preform a regular classification(with no outliers), take iris data set for example, it can not run:

import numpy as np
from sklearn.model_selection import train_test_split, ShuffleSplit
import pandas as pd
from GPS import GPSclassifier

columns = ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)','class']
df = pd.read_table("../datasets/data/iris/iris.data",delimiter=",",names=columns)
target = 'class'
features = columns[:-1]
df[target] = df[target].astype('category')
target_names = list(df[target].cat.categories)

'transform the class label to 1,2,3'
target_to_num = {target: i+1 for i, target in enumerate(target_names)}
df[target] = df[target].map(target_to_num)

X = df[features].values
y = df[target].values
K= np.unique(y).shape[0]

temp_X, test_X, temp_y, test_y = train_test_split(X, y, test_size=0.7, stratify=y)
train_X, calib_X, train_y, calib_y = train_test_split(temp_X, temp_y, test_size=0.5, stratify=temp_y)
calibLabelSet = np.hstack([calib_y.reshape(-1, 1), calib_X])

candC1 = [0.1, 1, 10]
candSigqtl = [25, 50, 75]

gps = GPSclassifier(alpha=0.05, K=3, calibSet=calibLabelSet, candC1=candC1, candSigqtl=candSigqtl)
gps.train(train_X, train_y)

**Comment by Zhou198** [comment]
Hi,

The paper targets the setting where training and test data come from different two distributions due to the emergence of the outlier class. That is why we incorporate unlabeled data (abuse marked as 0) containing outlier points from the test distribution with the labeled data from the training distribution, which refers to the lines of **newtrcaData** and **newtrcaL** in the Jupyter notebook. 

That being said, you still can use GPS if there is no outlier (the extreme case where the prior probability of the outlier class in test distribution is 0) in your data. The caveat is its OOD detection performance might not be good if later you want to work with the OOD detection task.

Going back to your code, the error arises due to your training part not including the unlabeled data (you can double-check with **train_y**), those points abuse marked as 0 as I mentioned in the 1st paragraph in this reply. To address this issue without modifying the code in the GPS.py, you may duplicate train_X but treat the duplicate part as "unlabeled data" (If you do not want to duplicate data, then you need to revise GPS.py). 

More concretely, the actual training data you use should be **dup_X = np.concatenate([train_X, train_X], 0)** with **dup_y = np.concatenate([train_y, train_y * 0])**. The same argument needs to be applied to your calibration data. Of course, alternatively, you can also replace the second **train_X** in **dup_X** with the unlabeled data from the test distribution (I assume both test and training distribution are the same in your setting with no outliers). 

I hope I address your question. Please let me know any other questions. Have fun.
